app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/StartController.scala

```diff
@@ -174,7 +175,7 @@ class StartController @Inject() (
     request: RequestWithSessionDataAndRetrievedData[AnyContent]
   ): Future[Result] =
     journeyStatus match {
-      case _: Subscribed =>
+      case _: Subscribed | _: Registering =>
```

Comment:
the lock should go on the repository layer, where `journeyStatus` comes from

Reply:
Not necessarily, because you want to send between status updates. IMO it should be pretty high.

Reply:
on second thought, i'm wondering if we need the lock at all, since updates are already done atomically with `findOneAndUpdate`

Reply:
We don't need locks on individual database operations - find and update is already atomic.

What you need a lock for is to create a transaction. Transactions are usually done in the application layer, not the database layer.

We want to implement the following bit of pseudocode, in the snippet above:

```scala
withLock {
  status <- getJourneyStatusFromCache
  if (status.type == SubmissionInProgress) {
    _ <- sendCreateAccountRequest
    _ <- sendEnrolmentRequest
    _ <- setJourneyStatusToComplete
  }
} match {
  case None => retryLock
  case Some(_) => Success
}
```

This ensures only one account creation request is done at any given time. Without it, the double submission, when performed quickly enough would still occur, as Ruth's noticed (two or more ifs can be true before you set the status journey to complete).

Reply:
I'm still not fully happy with it - if there's a crash when this chain of futures executes, the application will be left in an inconsistent state and rerunning the request may still create two accounts. But it's still miles ahead of what we have right now.